### PR TITLE
Clarify the patience in ReduceLROnPlateau

### LINKED ISSUE #119763: Learning rate reduction in ReduceLROnPlateau scheduler occurs one epoch later than expected due to patience parameter comparison
## 🐛 Describe the bug

The issue lies in the `ReduceLROnPlateau` scheduler's behavior when checking whether to reduce the learning rate based on the `patience` parameter. According to the PyTorch documentation, `patience` is described as the number of epochs with no improvement after which the learning rate will be reduced. For example, if `patience = 2`, then the scheduler should ignore the first 2 epochs with no improvement, and only decrease the learning rate after the 3rd epoch if there is still no improvement. 

However, in the current implementation, with `patience = 2`, the learning rate is actually reduced at the 4th epoch, which contradicts the expected behavior.

```python
if self.num_bad_epochs > self.patience:
    self._reduce_lr(epoch)
    self.cooldown_counter = self.cooldown
    self.num_bad_epochs = 0
```
 
 ## Possible solution

To address this issue, we can modify the condition in the `ReduceLROnPlateau` scheduler's `step` method to ensure that the learning rate reduction occurs after exactly `patience` epochs without improvement. 

```python
if self.num_bad_epochs >= self.patience:
    self._reduce_lr(epoch)
    self.cooldown_counter = self.cooldown
    self.num_bad_epochs = 0
   ```

### Versions

No Version needed. 

cc @svekars @brycebortree @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by lancerts**
Fixes #119763
@janeyx99

### CODE REVIEW

**Review by janeyx99 — DISMISSED** [review]
thanks for the fast turnaround!

**Comment by janeyx99** [comment]
Will land after verifying the docs preview.

**Comment by lancerts** [comment]
> Hey @lancerts after writing out [#119763 (comment)](https://github.com/pytorch/pytorch/issues/119763#issuecomment-1944423761), the current description may lead to more confusion. I think it would be helpful in the docs here to walk through the patience=0 case, preferably more concisely than what I delineated in the linked comment. Ping me for a review once that's ready!

Indeed, I walked through the case of patience = 2 case yesterday and found it confusing : P. I think patience =0 is the best example. Updated the message @janeyx99

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -965,11 +965,16 @@ class ReduceLROnPlateau(LRScheduler):
...
+        patience (int): The number of allowed epochs with no improvement after
+            which the learning rate will be reduced.
+            For example, with ``patience = 0``:
+            - 1st epoch: A baseline is established. As there's no previous baseline, the first epoch is always considered good.
+            - 2nd epoch: If this epoch performs worse than the baseline, num_bad_epochs increases by 1.
+            Since ``num_bad_epochs = 1 > patience = 0``, the learning rate will be reduced at the end of this epoch
+            and num_bad_epochs resets to 0.
+            - 3rd epoch onwards: If the epoch continues to perform worse than the baseline,
+            num_bad_epochs increases by 1. As ``num_bad_epochs = 1 > patience = 0``, the learning rate is reduced
+            and num_bad_epochs resets to 0 after each epoch.
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:977]:
```suggestion
            For example, with ``patience = 0``, the case of having no patience:
            - 1st epoch: A baseline is established. As there's no previous baseline, the first epoch is always considered good.
            - 2nd epoch: If this epoch performs worse than the baseline, we now have one untolerated epoch,  and 1 > 0, thus, the learning rate will be reduced at the end of this epoch.
            - 3rd epoch onwards: If the epoch continues to perform worse than the baseline, the learning rate will keep getting reduced at the end of the epoch. Otherwise, the lr will remain the same.
```

Comment (janeyx99) [reply torch/optim/lr_scheduler.py:977]: It's best to avoid including in-code details, since doc readers will not have that context

**Comment by lancerts** [comment]
@janeyx99 update the explanations based on the suggestion with some minor lint and touch. thanks

**Comment by janeyx99** [comment]
Okay let's see how this renders in our docs job. Thanks for the quick responses!

**Comment by lancerts** [comment]
> Okay let's see how this renders in our docs job. Thanks for the quick responses!

image
It seems the rendering doesnt follow the typical bullet point form for 1st,2nd and 3rd.
So maybe just make it normal paragraph? @janeyx99 
I feel the explanation is not OK but the `-` is a bit annoying in the paragraph. wdyt

**Comment by janeyx99** [comment]
Yea @lancerts could you rework it into a paragraph form? Might have to change some phrases around to make it flow.

**Comment by lancerts** [comment]
> Yea @lancerts could you rework it into a paragraph form? Might have to change some phrases around to make it flow.

@janeyx99 thanks for the suggestions, modified accordingly.

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -965,11 +965,17 @@ class ReduceLROnPlateau(LRScheduler):
...
-        patience (int): Number of epochs with no improvement after
-            which learning rate will be reduced. For example, if
-            `patience = 2`, then we will ignore the first 2 epochs
-            with no improvement, and will only decrease the LR after the
-            3rd epoch if the loss still hasn't improved then.
+        patience (int): The number of allowed epochs with no improvement after
+            which the learning rate will be reduced.
+            For example, consider the case of having no patience (`patience = 0`).
+            In the first epoch, a baseline is established and is always considered good as there's no previous baseline.
+            From the second epoch, if the performance is worse than the baseline,
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:972]:
```suggestion
            In the second epoch, if the performance is worse than the baseline,
```

**Review by janeyx99 — APPROVED** [review]
one more nit

**Comment by lancerts** [comment]
> one more nit

@janeyx99 thanks for the catch and commited
